File: PY/463-island-perimeter.py
def islandPerimeter(grid):
  count = 0
  grid_length, row_length = len(grid), len(grid[0])
  for i in range(grid_length):
    for j in range(row_length):
      if(grid[i][j] == 1):
        # value above
        if i == 0 or grid[i-1][j] == 0:
           count += 1 
        # value to the left:
        if j == 0 or grid[i][j-1] == 0:
           count += 1
        # value to the right
        if j == row_length - 1 or grid[i][j+1] == 0:
           count += 1
        # value below
        if i == grid_length - 1 or grid[i+1][j] == 0:
           count += 1
  
  return count

File: JS/463-island-perimeter.js
islandPerimeter = (grid) => {
  let total = 0
  const gridLength = grid.length
  const rowLength = grid[0].length 
  //double for loops ~10% faster than double forEach which is in turn ~10% faster than double map
  for(let i = 0; i <gridLength; i ++){
    for(let j = 0; j<rowLength; j++){
       if (grid[i][j] === 1){
       // value above
       (i - 1 === -1 || grid[i-1][j] === 0) && total++ 
       // value to the left:
       (j - 1 === -1 || grid[i][j-1] === 0) && total++
       //value to the right
       (grid[i][j+1] === undefined || grid[i][j+1] === 0) && total++
       //value below
       (grid[i+1] === undefined || grid[i+1][j] === 0) && total++
       }
    }
  }
 return total
}

console.log(islandPerimeter(grid))
